// openapi-to-postmanv2/dist/libV2/SpecificationCollectionSyncing/shared/constants.js
"use strict";
/**
 * Common constants used across SpecificationCollection modules
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SYNC_OPTIONS = exports.ALLOWED_AUTH_PARAM_KEYS_BY_TYPE = exports.AUTH_TYPES = exports.DEFAULT_RESPONSE_CODE_IN_OAS = exports.SPACE_COUNT = void 0;
/**
 * Number of spaces for JSON formatting
 */
exports.SPACE_COUNT = 2;
/**
 * Default response code used in OpenAPI specifications
 */
exports.DEFAULT_RESPONSE_CODE_IN_OAS = 'default';
/**
 * Authentication types supported in the system
 */
exports.AUTH_TYPES = {
    BASIC: 'basic',
    BEARER: 'bearer',
    DIGEST: 'digest',
    API_KEY: 'apikey',
    OAUTH2: 'oauth2'
};
/**
 * Allowed authentication parameter keys by authentication type
 */
exports.ALLOWED_AUTH_PARAM_KEYS_BY_TYPE = {
    [exports.AUTH_TYPES.API_KEY]: new Set(['in', 'key']),
    [exports.AUTH_TYPES.OAUTH2]: new Set([
        'scope',
        'authUrl',
        'authorizationUrl',
        'accessTokenUrl',
        'tokenUrl',
        'grant_type',
        'refreshUrl',
        'refresh_token_url'
    ])
};
exports.DEFAULT_SYNC_OPTIONS = {
    syncExamples: false
};
